renku-model/src/main/scala/io/renku/graph/model/entities/Plan.scala

```diff
@@ -61,21 +61,15 @@ object Plan {
     }
 
   implicit def decoder(implicit renkuUrl: RenkuUrl): JsonLDDecoder[Plan] =
-    JsonLDDecoder.entity(EntityTypes.of(prov / "Plan")) { cursor =>
-      cursor.getEntityTypes
-        .flatMap {
-          case ets if ets.contains(StepPlan.entityTypes) =>
-            StepPlan.decoder.apply(cursor)
-          case ets if ets.contains(CompositePlan.Ontology.entityTypes) =>
-            CompositePlan.decoder.apply(cursor)
-          case ets =>
-            Left(
-              DecodingFailure(
-                DecodingFailure.Reason.CustomReason(show"Expected composite plan or step plan, but got: $ets"),
-                cursor.jsonLD.toJson.hcursor
-              )
-            )
-        }
+    JsonLDDecoder.entity(EntityTypes of prov / "Plan") { cursor =>
+      lazy val noMatchFailure: Either[DecodingFailure, Plan] = cursor.getEntityTypes.flatMap { ets =>
+        DecodingFailure(
+          DecodingFailure.Reason.CustomReason(show"Cannot decode entity as Plan: $ets"),
+          cursor.jsonLD.toJson.hcursor
+        ).asLeft
+      }
+
+      StepPlan.decoder.apply(cursor) orElse CompositePlan.decoder.apply(cursor) orElse noMatchFailure
```

Comment:
Interesting! I _think_ I tried a similar solution at first but it failed some tests I created 🤔  (maybe the tests were not good 😄)

Reply:
Hm, don't really know what to say

Reply:
Sorry, it was just a thought in my head, nothing to answer :)